File: core/src/feedbacker_core/archive.py
"""Select sampled submissions from bulk downloads without opening the rest (#15, #17).

A sample can be spread across several sources: a main bulk zip, a zip from a
second submission point (e.g. late submissions), split zip parts, or single
downloaded files. Each sampled identifier is matched across all of them.

A file matches an identifier when the identifier appears in its file name as a
whole token, i.e. not as part of a longer run of letters or digits. Only file
names are read to decide; unmatched files are never opened.
"""

from __future__ import annotations

import re
import zipfile
from dataclasses import dataclass, field
from pathlib import Path, PurePosixPath

from feedbacker_core.structure import name_shape

SUPPORTED = {".docx", ".pdf"}


@dataclass(frozen=True)
class Member:
    source: Path  # the zip, or the single file itself
    name: str  # the member name inside the zip, or the file name

    @property
    def is_archive(self) -> bool:
        return self.source.suffix.lower() == ".zip"

    @property
    def file_name(self) -> str:
        return PurePosixPath(self.name).name

    @property
    def suffix(self) -> str:
        return Path(self.name).suffix.lower()

    def read(self) -> bytes:
        if not self.is_archive:
            return self.source.read_bytes()
        with zipfile.ZipFile(self.source) as z:
            return z.read(self.name)


@dataclass
class Selection:
    matched: dict[str, Member] = field(default_factory=dict)  # external_id -> member
    unmatched_ids: list[str] = field(default_factory=list)
    ambiguous: dict[str, list[Member]] = field(default_factory=dict)
    ignored_count: int = 0  # files not selected; never opened

    def problems(self, label=lambda external_id: external_id, sources=()) -> list[str]:
        """Describe problems without real file names, which may identify students.

        ``label`` names a sampled submission (e.g. by pseudonym); candidates are
        described by source position and name shape only.
        """
        index = {src: n for n, src in enumerate(sources, 1)}
        out = [f"no file found for {label(i)}" for i in self.unmatched_ids]
        for i, ms in self.ambiguous.items():
            where = ", ".join(
                f"source {index.get(m.source, '?')} ({name_shape(m.file_name)})" for m in ms
            )
            out.append(f"{label(i)} matches {len(ms)} files: {where}; resolve before importing")
        return out


def _token_pattern(external_id: str) -> re.Pattern[str]:
    return re.compile(rf"(?<![A-Za-z0-9]){re.escape(external_id)}(?![A-Za-z0-9])")


def list_members(source: Path) -> list[Member]:
    if source.suffix.lower() != ".zip":
        if not source.is_file():
            raise ValueError(f"source not found: {name_shape(source.name)}")
        return [Member(source=source, name=source.name)]
    try:
        with zipfile.ZipFile(source) as z:
            return [
                Member(source=source, name=i.filename)
                for i in z.infolist()
                if not i.is_dir()
                and not PurePosixPath(i.filename).name.startswith(".")
                and "__MACOSX" not in i.filename
            ]
    except (zipfile.BadZipFile, FileNotFoundError) as err:
        raise ValueError(
            f"not a readable zip archive ({name_shape(source.name)}): {type(err).__name__}"
        ) from None


def select_members(sources: Path | list[Path], external_ids: list[str]) -> Selection:
    if isinstance(sources, Path):
        sources = [sources]
    members = [m for s in sources for m in list_members(s)]
    selection = Selection()
    used: set[Member] = set()
    for external_id in external_ids:
        pattern = _token_pattern(external_id)
        hits = [m for m in members if pattern.search(m.file_name)]
        if not hits:
            selection.unmatched_ids.append(external_id)
        elif len(hits) > 1:
            selection.ambiguous[external_id] = hits
        else:
            selection.matched[external_id] = hits[0]
            used.add(hits[0])
    selection.ignored_count = len([m for m in members if m not in used])
    return selection

File: ui/src/core/archive.ts
/**
 * Select sampled submissions from bulk downloads without opening the rest: a
 * port of `core/src/feedbacker_core/archive.py` (#15, #17, #47).
 *
 * A sample can be spread across several sources: a main bulk zip, a zip from a
 * second submission point, split zip parts, or single downloaded files. Each
 * sampled identifier is matched across all of them. A file matches an
 * identifier when the identifier appears in its file name as a whole token.
 * Only names are read to decide: listing reads the zip's central directory,
 * and only the selected members are ever read (`readMember`).
 */

import { nameShape } from "./structure.ts";
import { listZip, readMember, type ByteSource, type ZipEntry } from "./zip.ts";

export class ArchiveError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ArchiveError";
  }
}

const baseName = (path: string) => path.slice(path.lastIndexOf("/") + 1);
const isZip = (source: ByteSource) => source.name.toLowerCase().endsWith(".zip");

export class Member {
  readonly source: ByteSource;
  /** The member name inside the zip, or the file name. */
  readonly name: string;
  readonly #entry: ZipEntry | null;

  constructor(source: ByteSource, name: string, entry: ZipEntry | null) {
    this.source = source;
    this.name = name;
    this.#entry = entry;
  }

  get isArchive(): boolean {
    return this.#entry !== null;
  }

  get fileName(): string {
    return baseName(this.name);
  }

  get suffix(): string {
    const name = this.fileName;
    const dot = name.lastIndexOf(".");
    return dot > 0 ? name.slice(dot).toLowerCase() : "";
  }

  /** Read this member: the only way a member's content is ever read. */
  read(): Promise<Uint8Array> {
    return this.#entry ? readMember(this.source, this.#entry) : this.source.read(0, this.source.size);
  }
}

export class Selection {
  matched = new Map<string, Member>(); // external_id -> member
  unmatchedIds: string[] = [];
  ambiguous = new Map<string, Member[]>();
  ignoredCount = 0; // files not selected; never opened

  /**
   * Describe problems without real file names, which may identify students.
   * `label` names a sampled submission (e.g. by pseudonym); candidates are
   * described by source position and name shape only.
   */
  problems(label: (externalId: string) => string = (id) => id, sources: ByteSource[] = []): string[] {
    const out = this.unmatchedIds.map((id) => `no file found for ${label(id)}`);
    for (const [id, members] of this.ambiguous) {
      const where = members
        .map((m) => {
          const index = sources.indexOf(m.source);
          return `source ${index >= 0 ? index + 1 : "?"} (${nameShape(m.fileName)})`;
        })
        .join(", ");
      out.push(`${label(id)} matches ${members.length} files: ${where}; resolve before importing`);
    }
    return out;
  }
}

const escapeRegExp = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
const tokenPattern = (externalId: string) => new RegExp(`(?<![A-Za-z0-9])${escapeRegExp(externalId)}(?![A-Za-z0-9])`);

export async function listMembers(source: ByteSource): Promise<Member[]> {
  if (!isZip(source)) return [new Member(source, source.name, null)];
  let entries: ZipEntry[];
  try {
    entries = await listZip(source);
  } catch (err) {
    throw new ArchiveError(`not a readable zip archive (${nameShape(source.name)}): ${(err as Error).name}`);
  }
  return entries
    .filter((e) => !e.isDirectory && !baseName(e.name).startsWith(".") && !e.name.includes("__MACOSX"))
    .map((e) => new Member(source, e.name, e));
}

export async function selectMembers(sources: ByteSource | ByteSource[], externalIds: string[]): Promise<Selection> {
  const list = Array.isArray(sources) ? sources : [sources];
  const members = (await Promise.all(list.map(listMembers))).flat();
  const selection = new Selection();
  const used = new Set<Member>();
  for (const id of externalIds) {
    const pattern = tokenPattern(id);
    const hits = members.filter((m) => pattern.test(m.fileName));
    if (!hits.length) selection.unmatchedIds.push(id);
    else if (hits.length > 1) selection.ambiguous.set(id, hits);
    else {
      selection.matched.set(id, hits[0]);
      used.add(hits[0]);
    }
  }
  selection.ignoredCount = members.filter((m) => !used.has(m)).length;
  return selection;
}
